component/publisher-client/src/main/java/org/wso2/am/analytics/publisher/retriever/MoesifKeyRetriever.java

```diff
@@ -0,0 +1,305 @@
+import com.moesif.api.MoesifAPIClient;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.wso2.am.analytics.publisher.exception.APICallException;
+import org.wso2.am.analytics.publisher.reporter.moesif.util.MoesifKeyEntry;
+import org.wso2.am.analytics.publisher.reporter.moesif.util.MoesifMicroserviceConstants;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.lang.reflect.Type;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.Base64;
+import java.util.Collection;
+import java.util.concurrent.ConcurrentHashMap;
+import javax.net.ssl.HttpsURLConnection;
+
+/**
+ * Responsible for calling the Moesif microservice and refresh/init entire internal map(organization, moesif key) or
+ * retrieving single moesif key.
+ * Also ease Moesif SDK client initiation by maintaining a map(moesif key, moesif sdk client {@link MoesifAPIClient}).
+ */
+public class MoesifKeyRetriever {
+    private static final Logger log = LoggerFactory.getLogger(MoesifKeyRetriever.class);
+    private static MoesifKeyRetriever moesifKeyRetriever;
+    private ConcurrentHashMap<String, String> orgIDMoesifKeyMap;
+
+    private ConcurrentHashMap<String, MoesifAPIClient> moesifKeyClientMap;
+    private String gaAuthUsername;
+    private char[] gaAuthPwd;
+
+    private MoesifKeyRetriever(String authUsername, String authPwd) {
+
+        this.gaAuthUsername = authUsername;
+        this.gaAuthPwd = authPwd.toCharArray();
+        orgIDMoesifKeyMap = new ConcurrentHashMap();
+        moesifKeyClientMap = new ConcurrentHashMap();
+    }
+
+    public static synchronized MoesifKeyRetriever getInstance(String authUsername, String authPwd) {
+        if (moesifKeyRetriever == null) {
+            return new MoesifKeyRetriever(authUsername, authPwd);
+        }
+        return moesifKeyRetriever;
+    }
+
+    /**
+     * Will initialize the empty orgID-MoesifKey map.
+     * Will refresh/refill the  orgID-MoesifKey map.
+     */
+    public void initOrRefreshOrgIDMoesifKeyMap() {
+        int attempts = MoesifMicroserviceConstants.NUM_RETRY_ATTEMPTS;
+        try {
+            callListResource();
+        } catch (IOException | APICallException ex) {
+            // TODO: Separate retry logic to a separate class.
+            log.error("First attempt of refreshing internal map failed,retrying.", ex);
+
+            while (attempts > 0) {
+                attempts--;
+                try {
+                    Thread.sleep(MoesifMicroserviceConstants.TIME_TO_WAIT);
+                } catch (InterruptedException e) {
+                    throw new RuntimeException(e);
+                }
+                try {
+                    callListResource();
+                } catch (IOException | APICallException e) {
+                    log.error("Retry attempt failed.", e);
```

Comment:
Let's improve this log also with the retry attempt number and what it is we are trying to refresh here.